<a href="https://colab.research.google.com/github/Vhakash/train-csv-data-prep/blob/main/Data_cleaning_and_prep.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Data cleaning and preprocessing using pandas.

import the pandas and numpy to perform the related operation


In [30]:
import pandas as pd
import numpy as np
df = pd.read_csv('train.csv')

check few rows of the data and the necessary information required for further steps.


In [31]:
df.head()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

## Identify missing values

### Subtask:
Determine which columns have missing values and how many.


**Reasoning**:
 I will first calculate the number of missing values for each column in the dataframe. Then, I will filter this result to only show the columns that actually contain missing values, making the output more concise and focused on the problem areas.



In [32]:
missing_values = df.isnull().sum()
missing_values = missing_values[missing_values > 0]
print(missing_values)

LotFrontage      259
Alley           1369
MasVnrType       872
MasVnrArea         8
BsmtQual          37
BsmtCond          37
BsmtExposure      38
BsmtFinType1      37
BsmtFinType2      38
Electrical         1
FireplaceQu      690
GarageType        81
GarageYrBlt       81
GarageFinish      81
GarageQual        81
GarageCond        81
PoolQC          1453
Fence           1179
MiscFeature     1406
dtype: int64


## Handle missing values

### Subtask:
Address missing values using appropriate techniques (e.g., imputation, removal).


**Reasoning**:
Drop columns with a high number of missing values and then impute missing values in the specified numerical and categorical columns.



In [33]:
df.drop(['Alley', 'PoolQC', 'Fence', 'MiscFeature'], axis = 1, inplace=True)

numerical_cols_to_impute = ['LotFrontage', 'MasVnrArea', 'GarageYrBlt']
for col in numerical_cols_to_impute:
    df[col].fillna(df[col].median(), inplace = True)

categorical_cols_to_impute = ['MasVnrType', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Electrical', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual', 'GarageCond']
for col in categorical_cols_to_impute:
    df[col].fillna(df[col].mode(), inplace = True)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 77 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1460 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   LotShape       1460 non-null   object 
 7   LandContour    1460 non-null   object 
 8   Utilities      1460 non-null   object 
 9   LotConfig      1460 non-null   object 
 10  LandSlope      1460 non-null   object 
 11  Neighborhood   1460 non-null   object 
 12  Condition1     1460 non-null   object 
 13  Condition2     1460 non-null   object 
 14  BldgType       1460 non-null   object 
 15  HouseStyle     1460 non-null   object 
 16  OverallQual    1460 non-null   int64  
 17  OverallCond    1460 non-null   int64  
 18  YearBuil

/tmp/ipython-input-2023521945.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(df[col].median(), inplace = True)
/tmp/ipython-input-2023521945.py:9: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try us

## Identify and handle duplicates

### Subtask:
Find and remove duplicate rows if any exist.


**Reasoning**:
Check for duplicate rows and remove them if found, then print the number of rows before and after removal.



In [34]:
rows_before = df.shape[0]
df.drop_duplicates(inplace=True)
rows_after = df.shape[0]
print(f"Number of rows before removing duplicates: {rows_before}")
print(f"Number of rows after removing duplicates: {rows_after}")

Number of rows before removing duplicates: 1460
Number of rows after removing duplicates: 1460


## Identify and handle outliers

### Subtask:
Detect and decide how to handle outliers in numerical columns.


**Reasoning**:
Select numerical columns, calculate IQR, define bounds, identify outliers, and replace them with the median in a single code block.



In [35]:
numerical_cols = df.select_dtypes(include=np.number).columns

for col in numerical_cols:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    median_value = df[col].median()
    df[col] = np.where((df[col] < lower_bound) | (df[col] > upper_bound), median_value, df[col])

df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
0  1.0        60.0       RL         65.0   8450.0   Pave      Reg         Lvl   
1  2.0        20.0       RL         80.0   9600.0   Pave      Reg         Lvl   
2  3.0        60.0       RL         68.0  11250.0   Pave      IR1         Lvl   
3  4.0        70.0       RL         60.0   9550.0   Pave      IR1         Lvl   
4  5.0        60.0       RL         84.0  14260.0   Pave      IR1         Lvl   

  Utilities LotConfig  ... EnclosedPorch 3SsnPorch ScreenPorch PoolArea  \
0    AllPub    Inside  ...           0.0       0.0         0.0      0.0   
1    AllPub       FR2  ...           0.0       0.0         0.0      0.0   
2    AllPub    Inside  ...           0.0       0.0         0.0      0.0   
3    AllPub    Corner  ...           0.0       0.0         0.0      0.0   
4    AllPub       FR2  ...           0.0       0.0         0.0      0.0   

  MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
0     0.0    2.0  2008.0        WD         Normal   208500.0  
1     0.0    5.0  2007.0        WD         Normal   181500.0  
2     0.0    9.0  2008.0        WD         Normal   223500.0  
3     0.0    2.0  2006.0        WD        Abnorml   140000.0  
4     0.0   12.0  2008.0        WD         Normal   250000.0  

[5 rows x 77 columns]

## Encode categorical variables

### Subtask:
Convert categorical variables into a numerical format suitable for analysis.


**Reasoning**:
Identify categorical columns, apply one-hot encoding, concatenate with the original dataframe, and display the head of the resulting dataframe.



In [9]:
categorical_cols = df.select_dtypes(include='object').columns
df_encoded = pd.get_dummies(df, columns=categorical_cols, drop_first=True)
display(df_encoded.head())

Id  MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
0  1.0        60.0         65.0   8450.0          7.0          5.0     2003.0   
1  2.0        20.0         80.0   9600.0          6.0          5.0     1976.0   
2  3.0        60.0         68.0  11250.0          7.0          5.0     2001.0   
3  4.0        70.0         60.0   9550.0          7.0          5.0     1915.0   
4  5.0        60.0         84.0  14260.0          8.0          5.0     2000.0   

   YearRemodAdd  MasVnrArea  BsmtFinSF1  ...  SaleType_ConLI  SaleType_ConLw  \
0        2003.0       196.0       706.0  ...           False           False   
1        1976.0         0.0       978.0  ...           False           False   
2        2002.0       162.0       486.0  ...           False           False   
3        1970.0         0.0       216.0  ...           False           False   
4        2000.0       350.0       655.0  ...           False           False   

   SaleType_New  SaleType_Oth  SaleType_WD  SaleCondition_AdjLand  \
0         False         False         True                  False   
1         False         False         True                  False   
2         False         False         True                  False   
3         False         False         True                  False   
4         False         False         True                  False   

   SaleCondition_Alloca  SaleCondition_Family  SaleCondition_Normal  \
0                 False                 False                  True   
1                 False                 False                  True   
2                 False                 False                  True   
3                 False                 False                 False   
4                 False                 False                  True   

   SaleCondition_Partial  
0                  False  
1                  False  
2                  False  
3                  False  
4                  False  

[5 rows x 237 columns]

## Scale numerical features

### Subtask:
Apply feature scaling to numerical columns if necessary.


**Reasoning**:
Identify numerical columns, initialize and fit the StandardScaler, transform the data, and replace the original numerical columns with the scaled values.



In [10]:
from sklearn.preprocessing import StandardScaler

numerical_cols = df_encoded.select_dtypes(include=np.number).columns
scaler = StandardScaler()
df_encoded[numerical_cols] = scaler.fit_transform(df_encoded[numerical_cols])
display(df_encoded.head())

Id  MSSubClass  LotFrontage   LotArea  OverallQual  OverallCond  \
0 -1.730865    0.412653    -0.274093 -0.266799     0.652641    -0.534267   
1 -1.728492   -1.008276     0.809612  0.101854    -0.077517    -0.534267   
2 -1.726120    0.412653    -0.057352  0.630791     0.652641    -0.534267   
3 -1.723747    0.767886    -0.635329  0.085825     0.652641    -0.534267   
4 -1.721374    0.412653     1.098601  1.595700     1.382798    -0.534267   

   YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  ...  SaleType_ConLI  \
0   1.060436      0.878668    1.290955    0.644729  ...           False   
1   0.145062     -0.429577   -0.595110    1.287800  ...           False   
2   0.992630      0.830215    0.963781    0.124597  ...           False   
3  -1.923005     -0.720298   -0.595110   -0.513745  ...           False   
4   0.958727      0.733308    2.772863    0.524153  ...           False   

   SaleType_ConLw  SaleType_New  SaleType_Oth  SaleType_WD  \
0           False         False         False         True   
1           False         False         False         True   
2           False         False         False         True   
3           False         False         False         True   
4           False         False         False         True   

   SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
0                  False                 False                 False   
1                  False                 False                 False   
2                  False                 False                 False   
3                  False                 False                 False   
4                  False                 False                 False   

   SaleCondition_Normal  SaleCondition_Partial  
0                  True                  False  
1                  True                  False  
2                  True                  False  
3                 False                  False  
4                  True                  False  

[5 rows x 237 columns]

## Review and validate

### Subtask:
Check the processed data for consistency and correctness.


**Reasoning**:
Display the data types, check for remaining missing values, display the head, and print the shape of the processed DataFrame to check for consistency and correctness after preprocessing.



In [11]:
print("Data types:")
print(df_encoded.dtypes)
print("\nRemaining missing values:")
print(df_encoded.isnull().sum().sum())
print("\nFirst few rows of processed DataFrame:")
display(df_encoded.head())
print("\nShape of processed DataFrame:")
print(df_encoded.shape)

Data types:
Id                       float64
MSSubClass               float64
LotFrontage              float64
LotArea                  float64
OverallQual              float64
                          ...   
SaleCondition_AdjLand       bool
SaleCondition_Alloca        bool
SaleCondition_Family        bool
SaleCondition_Normal        bool
SaleCondition_Partial       bool
Length: 237, dtype: object

Remaining missing values:
0

First few rows of processed DataFrame:


Id  MSSubClass  LotFrontage   LotArea  OverallQual  OverallCond  \
0 -1.730865    0.412653    -0.274093 -0.266799     0.652641    -0.534267   
1 -1.728492   -1.008276     0.809612  0.101854    -0.077517    -0.534267   
2 -1.726120    0.412653    -0.057352  0.630791     0.652641    -0.534267   
3 -1.723747    0.767886    -0.635329  0.085825     0.652641    -0.534267   
4 -1.721374    0.412653     1.098601  1.595700     1.382798    -0.534267   

   YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  ...  SaleType_ConLI  \
0   1.060436      0.878668    1.290955    0.644729  ...           False   
1   0.145062     -0.429577   -0.595110    1.287800  ...           False   
2   0.992630      0.830215    0.963781    0.124597  ...           False   
3  -1.923005     -0.720298   -0.595110   -0.513745  ...           False   
4   0.958727      0.733308    2.772863    0.524153  ...           False   

   SaleType_ConLw  SaleType_New  SaleType_Oth  SaleType_WD  \
0           False         False         False         True   
1           False         False         False         True   
2           False         False         False         True   
3           False         False         False         True   
4           False         False         False         True   

   SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
0                  False                 False                 False   
1                  False                 False                 False   
2                  False                 False                 False   
3                  False                 False                 False   
4                  False                 False                 False   

   SaleCondition_Normal  SaleCondition_Partial  
0                  True                  False  
1                  True                  False  
2                  True                  False  
3                 False                  False  
4                  True                  False  

[5 rows x 237 columns]


Shape of processed DataFrame:
(1460, 237)


## Summary:

### Data Analysis Key Findings

*   Initially, the dataframe had missing values in several columns, including 'Alley', 'PoolQC', 'Fence', 'MiscFeature', 'LotFrontage', 'MasVnrArea', 'GarageYrBlt', 'MasVnrType', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Electrical', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual', and 'GarageCond'.
*   Columns with a high number of missing values ('Alley', 'PoolQC', 'Fence', 'MiscFeature') were removed.
*   Missing values in numerical columns ('LotFrontage', 'MasVnrArea', 'GarageYrBlt') were imputed with the median of the respective columns.
*   Missing values in categorical columns ('MasVnrType', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Electrical', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual', 'GarageCond') were imputed with the mode of the respective columns.
*   No duplicate rows were found in the dataset; the number of rows remained at 1460 before and after the duplicate check.
*   Outliers in numerical columns were identified using the IQR method and replaced with the median value of the respective column.
*   Categorical variables were successfully converted into a numerical format using one-hot encoding, resulting in a significant increase in the number of columns.
*   Numerical features were scaled using `StandardScaler`.
*   After all preprocessing steps, the processed DataFrame (`df_encoded`) has 1460 rows and 237 columns, contains only `float64` and `bool` data types, and has no remaining missing values.

### Insights or Next Steps

*   The data is now cleaned and preprocessed, making it suitable for further analysis or machine learning model training.
*   Consider evaluating different imputation strategies or outlier handling methods to assess their impact on downstream analysis or model performance.
